### PR TITLE
Edgecore 4630/5835/7326/7816 API2.0 platform support

### PR DESCRIPTION
**Pull request by jeffhtgt**
#### Why I did it
Complete platform changes for AS4630-pe, AS5835-X, AS7326, AS7816 switch models (currently broken in master).

#### How I did it
Added proper support in sonic-buildimage repo in platform/broadcom/ and device/accton/.

#### How to verify it
Verified with the following commands:
- show platform fan
- show platform psustatus
- show platform syseeprom
- show platform temperature
- show interfaces status
- show interfaces transceiver eeprom --dom

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

#### Description for the changelog

Add API2.0 support/fixes for the following Edgecore/Accton switch models:
- as4630-54pe
- as5835-54x
- as7326-56x
- as7816-64x

#### Link to config_db schema for YANG module changes

N/A

#### A picture of a cute animal (not mandatory but encouraged)

image

### CODE REVIEW

**Comment by zhangyanzhao** [comment]
@jostar-yang please help to review this PR

**Comment by qiluo-msft** [comment]
@akenliu Please help review.

**Comment by prgeor** [comment]
@jeffhtgt please use SFP-refactor on your platform similar to this effor:- [https://github.com/Azure/sonic-buildimage/pull/10047](https://github.com/Azure/sonic-buildimage/pull/10047)

**Review by akenliu — COMMENTED** [review]
I think we should not remove the files about pddf.
These files are running for BRCM's PDDF platform.
Maybe, there is a method for selecting which platform (API-2.0 or PDDF) shall be used during compiling. (I am not sure.)

**Comment by jeffhtgt** [comment]
@prgeor This PR is for API2.0 support.  Can you point me to an example for sfp refactor using API2.0 only?

**Comment by prgeor** [comment]
> @prgeor This PR is for API2.0 support. Can you point me to an example for sfp refactor using API2.0 only?

[HLD Doc for SFP-refactor](https://github.com/Azure/SONiC/blob/master/doc/sfp-refactor/sfp-refactor.md)

Check these PRs:-
[PR9016](https://github.com/Azure/sonic-buildimage/pull/9016)
[PR10047](https://github.com/Azure/sonic-buildimage/pull/10047)

**Comment by jeffhtgt** [comment]
@prgeor I have not checked in any files yet but I am getting different SFP decode output after the refactoring.

Before refactoring:
Ethernet51: SFP EEPROM detected
        CalibrationType: Internally Calibrated
        Connector: LC
        EncodingCodes: 64B/66B
        ExtIdentOfTypeOfTransceiver: GBIC/SFP defined by twowire interface ID
        Length50um(UnitsOf10m): 8
        Length62.5um(UnitsOfm): 3
        LengthCable(UnitsOfm): 0
        LengthOM3(UnitsOf10m): 30
        LengthSMF(UnitsOf100m): 0
        LengthSMFkm-UnitsOfKm: 0
        NominalSignallingRate(UnitsOf100Mbd): 103
        RateIdentifier: Unspecified
        ReceivedPowerMeasurementType: Avg power
        TransceiverCodes:
                10GEthernetComplianceCode: 10G Base-SR
        TypeOfTransceiver: SFP/SFP+/SFP28
        VendorDataCode(YYYY-MM-DD Lot): 2021-02-08
        VendorName: PROLABS
        VendorOUI: 00-00-00
        VendorPN: ET5402-SR-C
        VendorRev: A1
        VendorSN: PRL84D053876
After:
Ethernet51: SFP EEPROM detected
        Application Advertisement: N/A
        Connector: LC
        Encoding: 64B/66B
        Extended Identifier: GBIC/SFP defined by two-wire interface ID
        Extended RateSelect Compliance: Unknown
        Identifier: SFP/SFP+/SFP28
        Length OM3(10m): 30.0
        Nominal Bit Rate(100Mbs): 103
        Specification compliance:
                10G Ethernet Compliance: Unknown
                ESCON Compliance: Unknown
                Ethernet Compliance: Unknown
                Fibre Channel Link Length: Unknown
                Fibre Channel Speed: Unknown
                Fibre Channel Transmission Media: Unknown
                Fibre Channel Transmitter Technology: Unknown
                Infiniband Compliance: Unknown
                SFP+CableTechnology: Unknown
                SONET Compliance Codes: Unknown
        Vendor Date Code(YYYY-MM-DD Lot): 2021-02-08   
        Vendor Name: PROLABS         
        Vendor OUI: 00-00-00
        Vendor PN: ET5402-SR-C     
        Vendor Rev: A1  
        Vendor SN: PRL84D053876 

Is this expected?

**Comment by prgeor** [comment]
> 10G Ethernet Compliance: Unknown
> ESCON Compliance: Unknown
> Ethernet Compliance: Unknown
> Fibre Channel Link Length: Unknown
> Fibre Channel Speed: Unknown
> Fibre Channel Transmission Media: Unknown
> Fibre Channel Transmitter Technology: Unknown
> Infiniband Compliance: Unknown
> SFP+CableTechnology: Unknown
> SONET Compliance Codes: Unknow

@jeffhtgt I can see few 'Unknown' for new fields. can you check the eeprom dump of the sfp? rest looks ok to me.

**Comment by jeffhtgt** [comment]
@prgeor Dump below.  Notice the "before" and "after" output of same SFP with SfpBase (before) and Optoe (after). 

root@t9847swta0003:/sys/bus/i2c/devices/20-0050# od -x eeprom 
0000000 0403 1007 0000 0000 0000 0600 0067 0000
0000020 0308 1e00 5250 4c4f 4241 2053 2020 2020
0000040 2020 2020 0000 0000 5445 3435 3230 532d
0000060 2d52 2043 2020 2020 3141 2020 5203 3400
0000100 1a00 0000 5250 384c 4434 3530 3833 3837
0000120 2020 2020 3132 3230 3830 2020 f068 3f03
0000140 0000 0000 0000 0000 0000 0000 0000 0000
*
0000200 ffff ffff ffff ffff ffff ffff ffff ffff
*
0000360 ffff ffff ffff ffff 3234 3633 3033 3534
0000400 005a 00f6 0055 00fb cc8d 0474 2887 447a
0000420 4c1d f401 7017 e803 e02e 3c06 1027 cb07
0000440 204e fa00 1027 f501 ffff ffff ffff ffff
0000460 ffff ffff ffff ffff 0000 0000 0000 0000
0000500 0000 0000 803f 0000 0000 0000 0001 0000
0000520 0001 0000 0001 0000 0001 0000 0000 4f00
0000540 eb22 9582 550c be15 0100 ffff ffff ff0a
0000560 4000 2000 4000 ffff ffff ffff ffff 00ff
0000600 0000 0000 0000 0000 0000 0000 0000 0000
*
0001000
root@t9847swta0003:/sys/bus/i2c/devices/20-0050#

**Comment by prgeor** [comment]
> @prgeor Dump below. Notice the "before" and "after" output of same SFP with SfpBase (before) and Optoe (after).
> 
> root@t9847swta0003:/sys/bus/i2c/devices/20-0050# od -x eeprom 0000000 0403 1007 0000 0000 0000 0600 0067 0000 0000020 0308 1e00 5250 4c4f 4241 2053 2020 2020 0000040 2020 2020 0000 0000 5445 3435 3230 532d 0000060 2d52 2043 2020 2020 3141 2020 5203 3400 0000100 1a00 0000 5250 384c 4434 3530 3833 3837 0000120 2020 2020 3132 3230 3830 2020 f068 3f03 0000140 0000 0000 0000 0000 0000 0000 0000 0000 * 0000200 ffff ffff ffff ffff ffff ffff ffff ffff * 0000360 ffff ffff ffff ffff 3234 3633 3033 3534 0000400 005a 00f6 0055 00fb cc8d 0474 2887 447a 0000420 4c1d f401 7017 e803 e02e 3c06 1027 cb07 0000440 204e fa00 1027 f501 ffff ffff ffff ffff 0000460 ffff ffff ffff ffff 0000 0000 0000 0000 0000500 0000 0000 803f 0000 0000 0000 0001 0000 0000520 0001 0000 0001 0000 0001 0000 0000 4f00 0000540 eb22 9582 550c be15 0100 ffff ffff ff0a 0000560 4000 2000 4000 ffff ffff ffff ffff 00ff 0000600 0000 0000 0000 0000 0000 0000 0000 0000 * 0001000 root@t9847swta0003:/sys/bus/i2c/devices/20-0050#

i don't see "before" and "after" labelled. What issue do you see?

**Comment by jeffhtgt** [comment]
@prgeor I was referring to comment #5.  The decode doesn't seem to be happening on some items with the refactored code.

**Comment by jeffhtgt** [comment]
@prgeor Let me be more specific regarding the output.  With SfpBase support these fields are displayed:
        Length50um(UnitsOf10m): 8
        Length62.5um(UnitsOfm): 3
        LengthCable(UnitsOfm): 0
        LengthSMF(UnitsOf100m): 0
        LengthSMFkm-UnitsOfKm: 0
        ReceivedPowerMeasurementType: Avg power
        TransceiverCodes:
                10GEthernetComplianceCode: 10G Base-SR

With refactoring, these are displayed instead of the above:
        Specification compliance:
                10G Ethernet Compliance: Unknown
                ESCON Compliance: Unknown
                Ethernet Compliance: Unknown
                Fibre Channel Link Length: Unknown
                Fibre Channel Speed: Unknown
                Fibre Channel Transmission Media: Unknown
                Fibre Channel Transmitter Technology: Unknown
                Infiniband Compliance: Unknown
                SFP+CableTechnology: Unknown
                SONET Compliance Codes: Unknown

If this is expected, I will move onto converting the other 3 switch platforms.

**Comment by prgeor** [comment]
> @prgeor Let me be more specific regarding the output. With SfpBase support these fields are displayed: Length50um(UnitsOf10m): 8 Length62.5um(UnitsOfm): 3 LengthCable(UnitsOfm): 0 LengthSMF(UnitsOf100m): 0 LengthSMFkm-UnitsOfKm: 0 ReceivedPowerMeasurementType: Avg power TransceiverCodes: 10GEthernetComplianceCode: 10G Base-SR
> 
> With refactoring, these are displayed instead of the above: Specification compliance: 10G Ethernet Compliance: Unknown ESCON Compliance: Unknown Ethernet Compliance: Unknown Fibre Channel Link Length: Unknown Fibre Channel Speed: Unknown Fibre Channel Transmission Media: Unknown Fibre Channel Transmitter Technology: Unknown Infiniband Compliance: Unknown SFP+CableTechnology: Unknown SONET Compliance Codes: Unknown
> 
> If this is expected, I will move onto converting the other 3 switch platforms.

@jeffhtgt please proceed with sfp-refactor. we can raise bug if any issue.

**Comment by jeffhtgt** [comment]
@prgeor See new commit.  All 4 platforms were tested with new SFP refactoring updates.

**Comment by prgeor** [comment]
@jeffhtgt could you resolve the merge conflict?

**Comment by jeffhtgt** [comment]
@prgeor Conflicts fixed.

**Comment by jeffhtgt** [comment]
@akenliu @prgeor Can this now be approved?  I believe I have made all requested changes.

File: device/accton/x86_64-accton_as4630_54pe-r0/sonic_platform/sfp.py [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+
+        reset_path="{}{}{}".format(CPLD_I2C_PATH , "module_reset_" , str(self.port_num))
+        val = self._api_helper.read_txt_file(reset_path)
+        
+        if val is not None:
+            return int(val, 10) == 1
+        else:        
+            return False
+
+    def get_rx_los(self):
```
Comment (prgeor) [inline device/accton/x86_64-accton_as4630_54pe-r0/sonic_platform/sfp.py:114]: why override this? the Sfp-refactor implements this in platform-independent way. get_tx_disable()  tx_disable_channel() set_power_override(), get_rx_los()? Why can't sfp-refactor access SFP eeprom in your platofrm?

Comment (jeffhtgt) [reply device/accton/x86_64-accton_as4630_54pe-r0/sonic_platform/sfp.py:114]: The functions I left in also use the platform CPLD.  I removed any functions that completely used eeprom only.

**Review by prgeor — APPROVED** [review]

**Pull request merged**